src/studio-app.js

```diff
@@ -200,58 +198,75 @@ export default class StudioApp {
     return imageUrls;
   }
 
-  // Private Methods
+  /**
+   * a flat list of all tags, associated with their DOM elements
+   */
+  get allTags() {
+    if (this._allTags) {
+      return this._allTags;
+    }
+
+    const flattenTags = (nestedTags = []) => {
+      return nestedTags.reduce((flattened, tag) => {
+        const { subtags } = tag;
+        return [...flattened, tag, ...flattenTags(subtags)];
+      }, []);
+    };
+
+    const tags = this._allTags = flattenTags(this.tags);
+    return tags;
+  }
 
   /**
-   * Pulls the tag list from `.mi-attributes` and associates tags with
-   * their DOM elements. It creates two properties, `this.tags` for the
-   * top-level tags in a tag hierarchy, and `this.allTags` for a flat
-   * list of all tags, including nested ones.
+   * a list of tags with nested subtags, associated with their DOM elements
    */
-  _setTags() {
-    this.tags = [];
-    this.allTags = [];
-    const attributesElement = document.querySelector('.mi-attributes');
+  get tags() {
+    if (this._tags) {
```

Comment:
It might be nice to have a pattern for memoization. _We could_ consider using a [memoization decorator](http://www.couchcoder.com/using-memoized-decorator-cache-computed-properties/)...but, if not, we could imitate the same behavior via the constructor:

```js
constructor() {
   // ...
   this.memoizer = new Memoizer();
}

get allTags() {
  return this.memoizer('allTags', () => {
    // ...
    return tags;
  });
}
```

Reply:
You might even be able to slim it down to `return this.memoizer(() => { ... }` if the memoizer looks at `Function.caller`?

Reply:
I've never heard of that! Looks really cool but also not usable? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/caller